### PR TITLE
Module forward reshape

### PR DESCRIPTION
**Pull request by kevinthesun**
Support module forward reshape. Resize executor if data batch shape changes.

### CODE REVIEW

File: python/mxnet/module/module.py [hunk 1]
```diff
@@ -539,6 +543,29 @@ def forward(self, data_batch, is_train=None):
             Default is ``None``, which means ``is_train`` takes the value of ``self.for_training``.
         """
         assert self.binded and self.params_initialized
+
+        # If start to inference, force rebind module.
+        if self._label_shapes and not data_batch.label:
+            raise RuntimeError("Currently module is bind with label while new data batch "
+                               "has no label. Are you trying to inference? Rebind module "
```
Comment (piiswrong) [inline python/mxnet/module/module.py]: "If you are trying to do inference, rebind module with 'force_rebind=True' and 'for_training=False'"

File: python/mxnet/module/module.py [hunk 2]
```diff
@@ -539,6 +543,29 @@ def forward(self, data_batch, is_train=None):
...
+                               "with 'force_rebind=True' and 'for_training=False'.")
+
+        for idx in range(len(self._data_shapes)):
+            curr_data_shape = self._data_shapes[idx].shape
+            new_data_shape = data_batch.data[idx].shape
```
Comment (piiswrong) [inline python/mxnet/module/module.py]: data_batch.data[idx][1]
instead of .shape

File: python/mxnet/module/module.py [hunk 3]
```diff
@@ -539,6 +543,29 @@ def forward(self, data_batch, is_train=None):
...
+            #Reshape if data shape changes.
+            if curr_data_shape != new_data_shape:
+                new_shape = data_batch.provide_data \
+                    if hasattr(data_batch, "provide_data") and data_batch.provide_data else \
+                    [(name, data.shape) for name, data in zip(self._data_names, data_batch.data)]
```
Comment (piiswrong) [inline python/mxnet/module/module.py]: also need to handle dtype and layout. Layout can be found from corresponding self._data_shapes

File: python/mxnet/module/module.py (hunk 2 above)
Comment (kevinthesun) [inline python/mxnet/module/module.py]: data_batch.data[idx][1] is a data sample from data_batch. Do we need data shape here?

Comment (piiswrong) [inline python/mxnet/module/module.py]: My mistake. Please ignore

**Comment by piiswrong** [comment]
moved to https://github.com/dmlc/mxnet/pull/6779

**Pull request closed without merging**